File: app/src/main/java/com/z/glad2see/ui/note_list/MainActivityPresenter.java
package com.z.glad2see.ui.note_list;

import android.util.Log;

import com.arellomobile.mvp.InjectViewState;
import com.arellomobile.mvp.MvpPresenter;
import com.z.glad2see.App;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.schedulers.Schedulers;

@InjectViewState
public class MainActivityPresenter extends MvpPresenter<MainActivityView> {

    public static final String TAG = "contact_list_debug";

    private CompositeDisposable compositeDisposable = new CompositeDisposable();

    public void onItemClicked(final long contactId) {
        getViewState().openContactEditorActivity(contactId);
    }

    @Override
    protected void onFirstViewAttach() {
        super.onFirstViewAttach();

        Log.d(TAG, "onFirstViewAttach: ");
        compositeDisposable.add(
                App.getContactManager().getAllContacts()
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe(
                                contacts -> {
                                    getViewState().showNotes();
                                    Log.d(TAG, "onFirstViewAttach: " + contacts.size());
                                },
                                throwable -> Log.d(TAG, "contact loading error", throwable)
                        )
        );
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        compositeDisposable.dispose();
    }
}

File: app/src/main/java/com/z/glad2see/ui/note_list/MainActivityPresenter.kt
package com.z.glad2see.ui.note_list

import com.arellomobile.mvp.InjectViewState
import com.arellomobile.mvp.MvpPresenter
import com.github.tamir7.contacts.Contact
import com.z.glad2see.App
import com.z.glad2see.addTo
import com.z.glad2see.disposeSafely
import com.z.glad2see.model.Note
import io.reactivex.Observable
import io.reactivex.Single
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers

@InjectViewState
class MainActivityPresenter : MvpPresenter<MainActivityView>() {

    private val subscriptions = CompositeDisposable()
    private var notesList = emptyList<NoteItemUi>()
    private val contactManager = App.getContactManager()

    fun onItemClicked(contactId: Long) {
        viewState.openContactEditorActivity(contactId)
    }

    override fun onFirstViewAttach() {
        super.onFirstViewAttach()
        updateNoteList()
    }

    fun updateNoteList() {
        getAllNotes()
            .flatMap { notes ->
                getContactsById(notes.map { it.contactId })
                    .flatMap { contacts ->
                        Single.fromCallable {
                            createNoteItems(contacts, notes)
                        }
                    }
            }
            .subscribeOn(Schedulers.newThread())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe(
                { itemList ->
                    processResults(itemList)
                },
                ::handleError
            )
            .addTo(subscriptions)
    }

    private fun getAllNotes(): Single<List<Note>> {
        return Single.fromCallable {
            App.getDatabase().notesDao.allNotes
        }
            .observeOn(Schedulers.newThread())
    }

    private fun getContactsById(ids: List<Long>): Single<List<Contact>> {
        return Single.fromCallable {
            val list = ArrayList<Contact>()
            ids.forEach { id ->
                list.add(contactManager.getContactById(id)[0])
            }
            list
        }
    }

    private fun createNoteItems(contacts: List<Contact>, notes: List<Note>): List<NoteItemUi> {
        return contacts.map { contact ->
            val note = notes.find { contact.id == it.contactId }
            val firstName = contact.givenName ?: ""
            val lastName = contact.familyName ?: ""
            NoteItemUi(contact.id, firstName, lastName, note?.textNote ?: "")
        }
    }

    private fun processResults(uiItems: List<NoteItemUi>) {
        notesList = uiItems
        viewState.showNotes(notesList)
    }

    override fun onDestroy() {
        super.onDestroy()
        subscriptions.disposeSafely()
    }

    private fun handleError(throwable: Throwable) {
        viewState.showState(throwable.message)
    }
}